File: libvbrief/issues.py
"""Validation issue and report models."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, List


@dataclass(frozen=True)
class Issue:
    """A single validation issue."""

    code: str
    path: str
    message: str
    severity: str


@dataclass
class ValidationReport:
    """Structured validation output used by strict and lenient modes."""

    errors: List[Issue] = field(default_factory=list)
    warnings: List[Issue] = field(default_factory=list)

    @property
    def is_valid(self) -> bool:
        """True when no errors were produced."""
        return not self.errors

    def add_error(self, code: str, path: str, message: str) -> None:
        """Append an error issue."""
        self.errors.append(Issue(code=code, path=path, message=message, severity="error"))

    def add_warning(self, code: str, path: str, message: str) -> None:
        """Append a warning issue."""
        self.warnings.append(Issue(code=code, path=path, message=message, severity="warning"))

    def extend(self, issues: Iterable[Issue]) -> None:
        """Append pre-built issues into report buckets by severity."""
        for issue in issues:
            if issue.severity == "warning":
                self.warnings.append(issue)
            else:
                self.errors.append(issue)

File: libxbrief-ts/src/issues.ts
/**
 * Validation issue and report models.
 */

export type IssueSeverity = "error" | "warning";

export interface IssueInit {
  code: string;
  path: string;
  message: string;
  severity: IssueSeverity;
}

/**
 * A single validation issue.
 */
export class Issue {
  public readonly code: string;
  public readonly path: string;
  public readonly message: string;
  public readonly severity: IssueSeverity;

  public constructor(init: IssueInit) {
    this.code = init.code;
    this.path = init.path;
    this.message = init.message;
    this.severity = init.severity;
  }
}

/**
 * Structured validation output used by strict and lenient modes.
 */
export class ValidationReport {
  public readonly errors: Issue[] = [];
  public readonly warnings: Issue[] = [];

  public get isValid(): boolean {
    return this.errors.length === 0;
  }

  public addError(code: string, path: string, message: string): void {
    this.errors.push(new Issue({ code, path, message, severity: "error" }));
  }

  public addWarning(code: string, path: string, message: string): void {
    this.warnings.push(new Issue({ code, path, message, severity: "warning" }));
  }

  public extend(issues: Iterable<Issue>): void {
    for (const issue of issues) {
      if (issue.severity === "warning") {
        this.warnings.push(issue);
      } else {
        this.errors.push(issue);
      }
    }
  }
}
